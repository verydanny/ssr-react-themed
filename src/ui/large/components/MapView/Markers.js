import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import { Markers as MapMarkers } from '@rentpath/react-ui-core'
import {
  emitMarkerClick,
  emitMarkerMouseOver,
  emitMarkerMouseOut,
  markerNormalIcon,
} from 'ui/shared/components/Markers'
import PinPopUp from './PinPopUp'

const MARKER_EVENTS = {
  onClick: emitMarkerClick,
  onMouseOver: emitMarkerMouseOver,
  onMouseOut: emitMarkerMouseOut,
}

export default class Markers extends PureComponent {
  static propTypes = {
    map: PropTypes.object,
    geojson: PropTypes.object,
    selectedListingId: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const selectedMarker = prevState.markers[nextProps.selectedListingId]

    if (selectedMarker !== prevState.selectedMarker) {
      return { selectedMarker }
    }

    return null
  }

  state = {
    selectedMarker: null,
    markers: {},
  }

  @autobind
  updateMarkers(markers) {
    this.setState({ markers })
  }

  markerFromFeature({ properties }) {
    return {
      icon: markerNormalIcon(properties),
      isPaid: properties.isPaid,
      title: null,
      zIndex: properties.isActive ? 2 : 1,
      ...MARKER_EVENTS,
    }
  }

  render() {
    const { selectedMarker } = this.state
    const { map, geojson } = this.props

    return (
      <Fragment>
        <MapMarkers
          map={map}
          geojson={geojson}
          marker={this.markerFromFeature}
          onMarkersReady={this.updateMarkers}
        />
        <PinPopUp
          map={map}
          selectedMarker={selectedMarker}
        />
      </Fragment>
    )
  }
}
