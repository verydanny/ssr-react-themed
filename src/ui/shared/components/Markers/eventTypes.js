//
//  Google Maps API Marker events
//
// These names match up to react-ui Markers naming.
// The event string should match Google's naming,
// to make it convenient for direct API usage when necessary
export const MARKER_ANIMATION_CHANGED = 'MARKER_animation_changed'
export const MARKER_CLICK = 'MARKER_click'
export const MARKER_CLICKABLE_CHANGED = 'MARKER_clickable_changed'
export const MARKER_CURSOR_CHANGED = 'MARKER_cursor_changed'
export const MARKER_DOUBLE_CLICK = 'MARKER_dblclick'
export const MARKER_DRAG = 'MARKER_drag'
export const MARKER_DRAG_END = 'MARKER_dragend'
export const MARKER_DRAGGABLE_CHANGED = 'MARKER_draggable_changed'
export const MARKER_DRAG_START = 'MARKER_dragstart'
export const MARKER_FLAT_CHANGED = 'MARKER_flat_changed'
export const MARKER_ICON_CHANGED = 'MARKER_icon_changed'
export const MARKER_MOUSE_DOWN = 'MARKER_mousedown'
export const MARKER_MOUSE_OUT = 'MARKER_mouseout'
export const MARKER_MOUSE_OVER = 'MARKER_mouseover'
export const MARKER_MOUSE_UP = 'MARKER_mouseup'
export const MARKER_POSITION_CHANGED = 'MARKER_position_changed'
export const MARKER_RIGHT_CLICK = 'MARKER_rightclick'
export const MARKER_SHAPE_CHANGED = 'MARKER_shape_changed'
export const MARKER_TITLE_CHANGED = 'MARKER_title_changed'
export const MARKER_VISIBLE_CHANGED = 'MARKER_visible_changed'
export const MARKER_Z_INDEX_CHANGED = 'MARKER_zindex_changed'
