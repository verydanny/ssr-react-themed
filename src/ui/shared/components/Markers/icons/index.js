export PinIcon from './map-pin.png'
export InactivePinIcon from './map-pin-inactive.png'
export SelectedPinIcon from './map-pin-selected.png'
export TeardropIcon from './map-teardrop.png'
export PdpPinLogoIcon from './pdp-pin-logo.png'
