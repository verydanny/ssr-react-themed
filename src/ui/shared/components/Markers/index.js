export {
  markerNormalIcon,
  markerSelectedIcon,
  PDP_LOGO_PIN,
} from './markerStyles'

export {
  emitMarkerAnimationChanged,
  emitMarkerClick,
  emitMarkerClickableChanged,
  emitMarkerCursorChanged,
  emitMarkerDoubleClick,
  emitMarkerDrag,
  emitMarkerDragEnd,
  emitMarkerDraggableChanged,
  emitMarkerDragStart,
  emitMarkerFlatChanged,
  emitMarkerIconChanged,
  emitMarkerMouseDown,
  emitMarkerMouseOut,
  emitMarkerMouseOver,
  emitMarkerMouseUp,
  emitMarkerPositionChanged,
  emitMarkerRightClick,
  emitMarkerShapeChanged,
  emitMarkerTitleChanged,
  emitMarkerVisibleChanged,
  emitMarkerZIndexChanged,
} from './events'

export {
  MARKER_ANIMATION_CHANGED,
  MARKER_CLICK,
  MARKER_CLICKABLE_CHANGED,
  MARKER_CURSOR_CHANGED,
  MARKER_DOUBLE_CLICK,
  MARKER_DRAG,
  MARKER_DRAG_END,
  MARKER_DRAGGABLE_CHANGED,
  MARKER_DRAG_START,
  MARKER_FLAT_CHANGED,
  MARKER_ICON_CHANGED,
  MARKER_MOUSE_DOWN,
  MARKER_MOUSE_OUT,
  MARKER_MOUSE_OVER,
  MARKER_MOUSE_UP,
  MARKER_POSITION_CHANGED,
  MARKER_RIGHT_CLICK,
  MARKER_SHAPE_CHANGED,
  MARKER_TITLE_CHANGED,
  MARKER_VISIBLE_CHANGED,
  MARKER_Z_INDEX_CHANGED,
} from './eventTypes'

export withMarkerEvents from './withMarkerEvents'
