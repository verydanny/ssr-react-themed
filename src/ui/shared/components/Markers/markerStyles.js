import {
  PinIcon,
  SelectedPinIcon,
  InactivePinIcon,
  PdpPinLogoIcon,
} from './icons'

const BIG_PIN = {
  url: PinIcon,
  anchor: { x: 8, y: 8 },
  scaledSize: { width: 16, height: 16 },
}

const SMALL_PIN = {
  url: PinIcon,
  anchor: { x: 5, y: 5 },
  scaledSize: { width: 11, height: 11 },
}

const INACTIVE_PIN = {
  url: InactivePinIcon,
  anchor: { x: 5, y: 5 },
  scaledSize: { width: 11, height: 11 },
}

const SELECTED_BIG_PIN = {
  url: SelectedPinIcon,
  anchor: { x: 8, y: 8 },
  scaledSize: { width: 16, height: 16 },
}

const SELECTED_SMALL_PIN = {
  url: SelectedPinIcon,
  anchor: { x: 5, y: 5 },
  scaledSize: { width: 11, height: 11 },
}

export const PDP_LOGO_PIN = {
  url: PdpPinLogoIcon,
  anchor: { x: 28, y: 32 },
  scaledSize: { width: 56, height: 64 },
}

export const markerNormalIcon = ({ isActive, isPaid }) => {
  if (!isActive) return INACTIVE_PIN
  return isPaid ? BIG_PIN : SMALL_PIN
}

export const markerSelectedIcon = isPaid => (isPaid ? SELECTED_BIG_PIN : SELECTED_SMALL_PIN)
