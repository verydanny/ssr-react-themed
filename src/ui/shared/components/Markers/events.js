import { Subject } from 'rxjs/Subject'
import * as types from './eventTypes'

export const eventSubject = new Subject()

const emit = type => (event, props, marker) => eventSubject.next({ type, event, props, marker })

//
//  Google Maps API Marker events
//  These come directly from Google's docs
//
export const emitMarkerAnimationChanged = emit(types.MARKER_ANIMATION_CHANGED)
export const emitMarkerClick = emit(types.MARKER_CLICK)
export const emitMarkerClickableChanged = emit(types.MARKER_CLICKABLE_CHANGED)
export const emitMarkerCursorChanged = emit(types.MARKER_CURSOR_CHANGED)
export const emitMarkerDoubleClick = emit(types.MARKER_DOUBLE_CLICK)
export const emitMarkerDrag = emit(types.MARKER_DRAG)
export const emitMarkerDragEnd = emit(types.MARKER_DRAG_END)
export const emitMarkerDraggableChanged = emit(types.MARKER_DRAGGABLE_CHANGED)
export const emitMarkerDragStart = emit(types.MARKER_DRAG_START)
export const emitMarkerFlatChanged = emit(types.MARKER_FLAT_CHANGED)
export const emitMarkerIconChanged = emit(types.MARKER_ICON_CHANGED)
export const emitMarkerMouseDown = emit(types.MARKER_MOUSE_DOWN)
export const emitMarkerMouseOut = emit(types.MARKER_MOUSE_OUT)
export const emitMarkerMouseOver = emit(types.MARKER_MOUSE_OVER)
export const emitMarkerMouseUp = emit(types.MARKER_MOUSE_UP)
export const emitMarkerPositionChanged = emit(types.MARKER_POSITION_CHANGED)
export const emitMarkerRightClick = emit(types.MARKER_RIGHT_CLICK)
export const emitMarkerShapeChanged = emit(types.MARKER_SHAPE_CHANGED)
export const emitMarkerTitleChanged = emit(types.MARKER_TITLE_CHANGED)
export const emitMarkerVisibleChanged = emit(types.MARKER_VISIBLE_CHANGED)
export const emitMarkerZIndexChanged = emit(types.MARKER_Z_INDEX_CHANGED)
