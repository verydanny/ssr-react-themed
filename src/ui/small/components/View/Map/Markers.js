import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import { Markers as MapMarkers } from '@rentpath/react-ui-core'
import { emitMarkerClick, markerNormalIcon } from 'ui/shared/components/Markers'
import { MarkerSelector } from 'ui/small/components/Map'

const MARKER_EVENTS = {
  onClick: emitMarkerClick,
}

export default class Markers extends PureComponent {
  static propTypes = {
    map: PropTypes.object,
    geojson: PropTypes.object,
    selectedListingId: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
    handleCardDrawer: PropTypes.func,
  }

  static defaultProps = {
    handleCardDrawer: () => {},
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const selectedMarker = prevState.markers[nextProps.selectedListingId]

    if (selectedMarker !== prevState.selectedMarker) {
      return { selectedMarker }
    }

    return null
  }

  state = {
    selectedMarker: null,
    markers: {},
  }

  @autobind
  updateMarkers(markers) {
    this.setState({ markers })
  }

  markerFromFeature({ properties }) {
    return {
      icon: markerNormalIcon(properties),
      isPaid: properties.isPaid,
      zIndex: properties.isActive ? 2 : 1,
      title: null,
      ...MARKER_EVENTS,
    }
  }

  render() {
    const { map, geojson, handleCardDrawer } = this.props
    const { selectedMarker } = this.state

    return (
      <Fragment>
        <MapMarkers
          map={map}
          geojson={geojson}
          marker={this.markerFromFeature}
          onMarkersReady={this.updateMarkers}
        />
        <MarkerSelector
          map={map}
          selectedMarker={selectedMarker}
          handleCardDrawer={handleCardDrawer}
        />
      </Fragment>
    )
  }
}
