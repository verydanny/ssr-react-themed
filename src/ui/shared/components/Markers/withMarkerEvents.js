import React, { PureComponent } from 'react'
import { Subject } from 'rxjs/Subject'
import { takeUntil } from 'rxjs/operators'
import { eventSubject } from './events'

export default function(BaseComponent) {
  class Container extends PureComponent {
    static displayName = `withMarkerEvents(${BaseComponent.displayName || BaseComponent.name || 'Component'})`

    constructor(props) {
      super(props)
      this.destroy = new Subject()
      this.observable = eventSubject.pipe(takeUntil(this.destroy))
    }

    componentWillUnmount() {
      this.destroy.next(true)
      this.destroy.unsubscribe()
    }

    render() {
      return (
        <BaseComponent
          markerObservable={this.observable}
          {...this.props}
        />
      )
    }
  }

  return Container
}
