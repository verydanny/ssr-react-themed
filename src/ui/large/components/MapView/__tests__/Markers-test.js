import React from 'react'
import { shallow } from 'enzyme'
import 'google-maps-mock'
import Markers from '../Markers'

const testListingId = '1234567'

const testMap = new window.google.maps.Map()

const testMarker = new window.google.maps.Marker({ map: testMap })
testMarker.id = testListingId

const testMarkers = {
  [testListingId]: testMarker,
}

const testFeature = {
  type: 'Feature',
  geometry: {
    type: 'Point',
    coordinates: [
      -84.38781,
      33.786645,
    ],
  },
  properties: {
    id: '2034989',
    name: 'Atlantic House',
    isPaid: true,
    isActive: true,
  },
}

const inactiveFeature = {
  type: 'Feature',
  geometry: {
    type: 'Point',
    coordinates: [
      -84.38781,
      33.786645,
    ],
  },
  properties: {
    id: '2034989',
    name: 'Atlantic House',
    isPaid: false,
    isActive: false,
  },
}

describe('ui/large/components/MapView/Markers', () => {
  it('sets the selectedMarker state when selected listing id changes', () => {
    const wrapper = shallow(<Markers geojson={{}} />)
    wrapper.instance().updateMarkers(testMarkers)
    wrapper.setProps({ selectedListingId: testListingId })
    expect(wrapper.state('selectedMarker')).toBe(testMarker)
  })

  it('sets correct marker data for active listings', () => {
    const wrapper = shallow(<Markers geojson={{}} />)
    expect(wrapper.instance().markerFromFeature(testFeature)).toMatchSnapshot()
  })

  it('sets the correct marker data for inactive listings', () => {
    const wrapper = shallow(<Markers geojson={{}} />)
    expect(wrapper.instance().markerFromFeature(inactiveFeature)).toMatchSnapshot()
  })
})
