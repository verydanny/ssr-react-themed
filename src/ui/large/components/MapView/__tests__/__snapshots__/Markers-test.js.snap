// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/large/components/MapView/Markers sets correct marker data for active listings 1`] = `
Object {
  "icon": Object {
    "anchor": Object {
      "x": 8,
      "y": 8,
    },
    "scaledSize": Object {
      "height": 16,
      "width": 16,
    },
    "url": "",
  },
  "isPaid": true,
  "onClick": [Function],
  "onMouseOut": [Function],
  "onMouseOver": [Function],
  "title": null,
  "zIndex": 2,
}
`;

exports[`ui/large/components/MapView/Markers sets the correct marker data for inactive listings 1`] = `
Object {
  "icon": Object {
    "anchor": Object {
      "x": 5,
      "y": 5,
    },
    "scaledSize": Object {
      "height": 11,
      "width": 11,
    },
    "url": "",
  },
  "isPaid": false,
  "onClick": [Function],
  "onMouseOut": [Function],
  "onMouseOver": [Function],
  "title": null,
  "zIndex": 1,
}
`;
